lib/CL/devices/level0/level0-driver.cc

```diff
@@ -0,0 +1,2239 @@
+  HostDeviceRate = HostDevRate;
+  HostTimingStart = HostTimeStart;
+  DeviceTimingStart = DeviceTimeStart;
+  DeviceMaxWGSizes = DeviceMaxWGs;
+  Thread = std::thread(&Level0Queue::runThread, this);
+}
+
+Level0Queue::~Level0Queue() {
+  if (Thread.joinable())
+    Thread.join();
+  if (CmdListH != nullptr) {
+    zeCommandListDestroy(CmdListH);
+  }
+  if (QueueH != nullptr) {
+    zeCommandQueueDestroy(QueueH);
+  }
+}
+
+// ***********************************************************************
+// ***********************************************************************
+
+bool Level0QueueGroup::init(unsigned Ordinal, unsigned Count,
+                            ze_context_handle_t ContextH,
+                            ze_device_handle_t DeviceH, uint64_t *Buffer,
+                            uint64_t HostTimingStart,
+                            uint64_t DeviceTimingStart, double HostDeviceRate,
+                            uint32_t_3 DeviceMaxWGs) {
+
+  ThreadExitRequested = false;
+
+  std::vector<ze_command_queue_handle_t> QHandles;
+  std::vector<ze_command_list_handle_t> LHandles;
+  assert(Count > 0);
+  QHandles.resize(Count);
+  LHandles.resize(Count);
+  ze_result_t r;
+  ze_command_queue_handle_t Queue;
+  ze_command_list_handle_t CmdList;
+
+  ze_command_queue_desc_t cmdQueueDesc = {ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC,
+                                          nullptr,
+                                          Ordinal,
+                                          0, // index
+                                          0, // flags
+                                          ZE_COMMAND_QUEUE_MODE_SYNCHRONOUS,
+                                          ZE_COMMAND_QUEUE_PRIORITY_NORMAL};
+
+  ze_command_list_desc_t cmdListDesc = {
+      ZE_STRUCTURE_TYPE_COMMAND_LIST_DESC, nullptr, Ordinal,
+      ZE_COMMAND_LIST_FLAG_RELAXED_ORDERING |
+          ZE_COMMAND_LIST_FLAG_MAXIMIZE_THROUGHPUT};
+
+  for (unsigned i = 0; i < Count; ++i) {
+    cmdQueueDesc.index = i;
+    r = zeCommandQueueCreate(ContextH, DeviceH, &cmdQueueDesc, &Queue);
+    LEVEL0_CHECK_RET(false, r);
+    r = zeCommandListCreate(ContextH, DeviceH, &cmdListDesc, &CmdList);
+    LEVEL0_CHECK_RET(false, r);
+    QHandles[i] = Queue;
+    LHandles[i] = CmdList;
+  }
+
+  for (unsigned i = 0; i < Count; ++i) {
+    Queues.emplace_back(new Level0Queue(
+        this, QHandles[i], LHandles[i], &Buffer[i * 8], HostDeviceRate,
+        HostTimingStart, DeviceTimingStart, DeviceMaxWGs));
+  }
+
+  return true;
+}
+
+Level0QueueGroup::~Level0QueueGroup() {
+  std::unique_lock<std::mutex> Lock(Mutex);
+  ThreadExitRequested = true;
+  Cond.notify_all();
+  Lock.unlock();
+  Queues.clear();
+}
+
+void Level0QueueGroup::pushWork(_cl_command_node *Command) {
+  std::lock_guard<std::mutex> Lock(Mutex);
+  WorkQueue.push(Command);
+  Cond.notify_one();
+}
+
+_cl_command_node *Level0QueueGroup::getWorkOrWait(bool &ShouldExit) {
+  _cl_command_node *cmd = nullptr;
```

Comment:
Command (capital first letter due to C++ style).

Reply:
fixed